python/desc/twinkles/validate_ic.py

```diff
@@ -0,0 +1,849 @@
+
+        lens_gals = df_galaxy.iloc[np.unique(lens_gal_locs)]
+        lens_gals = lens_gals.reset_index(drop=True)
+
+        return lens_gals
+
+    def process_sne_lenses(self, df_galaxy):
+
+        """
+        These must be done differently since the lens galaxy for SNe
+        will not have associated images in every instance catalog.
+        """
+
+        galtile_list = np.genfromtxt(os.path.join(os.environ['TWINKLES_DIR'],
+                                                  'data', 'dc2_sne_cache.csv'),
+                                     delimiter=',', names=True, dtype=np.int)
+
+        lens_gal_locs = []
+        twinkles_sys = []
+        for gal_row in galtile_list:
+            bulge_id = np.left_shift(gal_row['galtileid'], 10) + 97
+            match = np.where(df_galaxy['uniqueId'] == bulge_id)[0]
+            if len(match) > 1:
+                raise ValueError('Multiple galaxies with same id.')
+            elif len(match) == 1:
+                lens_gal_locs.append(match[0])
+                twinkles_sys.append(gal_row['twinkles_system'])
+
+        lens_gals = df_galaxy.iloc[lens_gal_locs]
+        lens_gals = lens_gals.reset_index(drop=True)
+        lens_gals['twinkles_system'] = twinkles_sys
+
+        return lens_gals
+                
+
+    def process_sprinkled_sne(self, df_sne):
+
+        galtileids = []
+        twinkles_system = []
+        twinkles_im_num = []
+        galtile_list = np.genfromtxt(os.path.join(os.environ['TWINKLES_DIR'],
+                                                  'data', 'dc2_sne_cache.csv'),
+                                     delimiter=',', names=True, dtype=np.int)
+        
+        i=0
+        keep_idx = []
+        for sne_idx, sne_row in df_sne.iterrows():
+            if sne_row['sedFilepath'].startswith('Dynamic'):
+                sne_id = sne_row['uniqueId']
+                twinkles_ids = np.right_shift(sne_id-117, 10)
+                galtileid = int(twinkles_ids/10000)
+                if galtileid in galtile_list['galtileid']:
+                    keep_idx.append(i)
+                    galtileids.append(galtileid)
+                    twinkles_num = np.int(str(twinkles_ids)[-4:])
+                    twinkles_system.append(twinkles_num // 4)
+                    twinkles_im_num.append(twinkles_num % 4)
+                
+            i+=1
+
+        galtileids = np.array(galtileids, dtype=np.int)
+        sprinkled_sne = df_sne.iloc[np.array(keep_idx)]
+        sprinkled_sne = sprinkled_sne.reset_index(drop=True)
+        sprinkled_sne['galaxy_id'] = galtileids
+        sprinkled_sne['twinkles_system'] = twinkles_system
+        sprinkled_sne['image_number'] = twinkles_im_num
+        sprinkled_sne['lens_galaxy_uID'] = np.left_shift(np.array(galtileids, dtype=np.int), 10) + 97
+        sprinkled_sne['host_galaxy_bulge_uID'] = np.array(np.left_shift(galtileids*10000
+                                                    + np.array(twinkles_system, dtype=np.int)*4 +
```

Comment:
Same comment as above about not casting `twinkles_system` and `twinkles_im_num` as numpy arrays more than once.